src/server/nuclearnet/tests/fake_nuclearnet_client.tests.ts

```diff
@@ -0,0 +1,192 @@
+
+    const disconnectEve = eve.connect({ name: 'eve' })
+    disconnectEve()
+
+    expect(bobOnLeave).toHaveBeenCalledTimes(1)
+    expect(bobOnLeave).toHaveBeenLastCalledWith(expect.objectContaining({ name: 'eve' }))
+
+    expect(aliceOnLeave).toHaveBeenCalledTimes(1)
+    expect(aliceOnLeave).toHaveBeenLastCalledWith(expect.objectContaining({ name: 'eve' }))
+  })
+
+  it('does not receive leave events from other clients after unsubscribing', () => {
+    bob.connect({ name: 'bob' })
+    alice.connect({ name: 'alice' })
+
+    const bobOnLeave = jest.fn()
+    const bobOffLeave = bob.onLeave(bobOnLeave)
+    bobOffLeave()
+
+    const aliceOnLeave = jest.fn()
+    alice.onLeave(aliceOnLeave)
+
+    const disconnectEve = eve.connect({ name: 'eve' })
+    disconnectEve()
+
+    expect(bobOnLeave).not.toHaveBeenCalled()
+
+    expect(aliceOnLeave).toHaveBeenCalledTimes(1)
+    expect(aliceOnLeave).toHaveBeenLastCalledWith(expect.objectContaining({ name: 'eve' }))
+  })
+
+  it('receives join events for all other connected clients on connect', () => {
+    bob.connect({ name: 'bob' })
+    alice.connect({ name: 'alice' })
+
+    const onJoin = jest.fn()
+    eve.onJoin(onJoin)
+    eve.connect({ name: 'eve' })
+
+    expect(onJoin).toHaveBeenCalledTimes(3)
+    expect(onJoin).toHaveBeenCalledWith(expect.objectContaining({ name: 'bob' }))
+    expect(onJoin).toHaveBeenCalledWith(expect.objectContaining({ name: 'alice' }))
+    expect(onJoin).toHaveBeenCalledWith(expect.objectContaining({ name: 'eve' }))
+  })
+
+  it('receives own messages', () => {
+    bob.connect({ name: 'bob' })
+
+    const onSensors = jest.fn()
+    bob.on('sensors', onSensors)
+
+    const payload = new Buffer(8)
+    bob.send({ type: 'sensors', payload })
+
+    expect(onSensors).toHaveBeenCalledTimes(1)
+    expect(onSensors).toHaveBeenLastCalledWith({ payload, peer: expect.objectContaining({ name: 'bob' }) })
+  })
+
+  it('receives messages sent from other clients', () => {
+    bob.connect({ name: 'bob' })
+    alice.connect({ name: 'alice' })
+    eve.connect({ name: 'eve' })
+
+    const bobOnSensors = jest.fn()
+    bob.on('sensors', bobOnSensors)
+
+    const aliceOnSensors = jest.fn()
+    alice.on('sensors', aliceOnSensors)
+
+    const eveOnSensors = jest.fn()
+    eve.on('sensors', eveOnSensors)
+
+    const payload = new Buffer(8)
+    eve.send({ type: 'sensors', payload })
+
+    expect(bobOnSensors).toHaveBeenCalledTimes(1)
+    expect(bobOnSensors).toHaveBeenLastCalledWith({ payload, peer: expect.objectContaining({ name: 'eve' }) })
+
+    expect(aliceOnSensors).toHaveBeenCalledTimes(1)
+    expect(aliceOnSensors).toHaveBeenLastCalledWith({ payload, peer: expect.objectContaining({ name: 'eve' }) })
+
+    expect(eveOnSensors).toHaveBeenCalledTimes(1)
+    expect(aliceOnSensors).toHaveBeenLastCalledWith({ payload, peer: expect.objectContaining({ name: 'eve' }) })
+  })
+
+  it('only receives targetted messages if they are the target', () => {
```

Comment:
You should also test that two clients with the same name will both receive messages sent to a target